relayer/workers/worker.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2021 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package workers
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type Worker interface {
+	Name() string
+
+	SetUp() error
+	TearDown()
+
+	Start(ctx context.Context, eg *errgroup.Group) error
+}
+
+
+type WorkerPool struct {
+	workers []Worker
+}
+
+func NewWorkerPool(workers []Worker) *WorkerPool {
+	return &WorkerPool{
+		workers,
+	}
+}
+
+func (wp *WorkerPool) runWorker(ctx context.Context, worker Worker) error {
+	// Ensure we always clean up after ourselves
+	defer worker.TearDown()
+
+	err := worker.SetUp()
+	if err != nil {
+		return err
+	}
+
+	childEg, childCtx := errgroup.WithContext(ctx)
+	err = worker.Start(childCtx, childEg)
+	if err != nil {
+		return err
+	}
+
+	return childEg.Wait()
+}
+
+func (wp *WorkerPool) Run() error {
```

Comment:
nit: Would prefer `Start` to be consistent with `worker.Start`

Reply:
I named it differently because the behaviour is different:
* `Start`: starts goroutines and returns
* `Run`: starts goroutines and blocks until `eg.Wait` returns and the worker is done

Yay or nay?

Reply:
:+1: